projects/packages/ip/tests/php/bootstrap.php

```diff
@@ -9,3 +9,4 @@
  * Include the composer autoloader.
  */
 require_once __DIR__ . '/../../vendor/autoload.php';
+require_once __DIR__ . '/functions-wordpress.php';
```

Comment:
Instead of mocking WordPress functions, what do you think about relying on `Brain\Monkey` directly in your tests, like so?

```diff
diff --git a/projects/packages/ip/composer.json b/projects/packages/ip/composer.json
index 3c8783a659..a5854dbf35 100644
--- a/projects/packages/ip/composer.json
+++ b/projects/packages/ip/composer.json
@@ -5,6 +5,7 @@
 	"license": "GPL-2.0-or-later",
 	"require": {},
 	"require-dev": {
+		"brain/monkey": "2.6.1",
 		"yoast/phpunit-polyfills": "1.0.4",
 		"automattic/jetpack-changelogger": "@dev"
 	},
diff --git a/projects/packages/ip/tests/php/bootstrap.php b/projects/packages/ip/tests/php/bootstrap.php
index dc812172a4..fa1b9f56ea 100644
--- a/projects/packages/ip/tests/php/bootstrap.php
+++ b/projects/packages/ip/tests/php/bootstrap.php
@@ -9,4 +9,3 @@
  * Include the composer autoloader.
  */
 require_once __DIR__ . '/../../vendor/autoload.php';
-require_once __DIR__ . '/functions-wordpress.php';
diff --git a/projects/packages/ip/tests/php/functions-wordpress.php b/projects/packages/ip/tests/php/functions-wordpress.php
deleted file mode 100644
index f4999b9a92..0000000000
--- a/projects/packages/ip/tests/php/functions-wordpress.php
+++ /dev/null
@@ -1,65 +0,0 @@
-<?php
-/**
- * Functions that mock WordPress core functionality for testing purposes.
- *
- * @package automattic/jetpack-ip
- */
-
-if ( ! function_exists( 'get_option' ) ) {
-	/**
-	 * A drop-in for a WordPress core function.
-	 *
-	 * Note that the `$test_options` global can be used to set options to be returned via this function.
-	 *
-	 * @param string $option  Name of the option to retrieve. Expected to not be SQL-escaped.
-	 * @param mixed  $default Default value to return if the option does not exist.
-	 * @return mixed Value set for the option.
-	 */
-	function get_option( $option, $default = false ) {
-		global $test_options;
-
-		if ( isset( $test_options[ $option ] ) ) {
-			return $test_options[ $option ];
-		}
-
-		return $default;
-	}
-}
-
-if ( ! function_exists( 'get_site_option' ) ) {
-	/**
-	 * A drop-in for a WordPress core function.
-	 *
-	 * Note that the `$test_options` global can be used to set options to be returned via this function.
-	 *
-	 * @param string $option  Name of the option to retrieve. Expected to not be SQL-escaped.
-	 * @param mixed  $default Default value to return if the option does not exist.
-	 * @return mixed Value set for the option.
-	 */
-	function get_site_option( $option, $default = false ) {
-		return get_option( $option, $default );
-	}
-}
-
-/**
- * Adds an option to be used in tests.
- *
- * @param string $option The option to set.
- * @param mixed  $value The value to set to the option.
- */
-function add_test_option( $option, $value ) {
-	global $test_options;
-	$test_options[ $option ] = $value;
-}
-
-if ( ! function_exists( 'wp_unslash' ) ) {
-	/**
-	 * A drop-in for a WordPress core function.
-	 *
-	 * @param string|array $value String or array of data to unslash.
-	 * @return string|array Unslashed `$value`.
-	 */
-	function wp_unslash( $value ) {
-		return $value;
-	}
-}
diff --git a/projects/packages/ip/tests/php/test-ip.php b/projects/packages/ip/tests/php/test-ip.php
index f5fc1509c4..a977ac9200 100644
--- a/projects/packages/ip/tests/php/test-ip.php
+++ b/projects/packages/ip/tests/php/test-ip.php
@@ -6,11 +6,32 @@
  */
 
 use Automattic\Jetpack\IP\IP;
+use Brain\Monkey;
+use Brain\Monkey\Functions;
 
 /**
  * IP class test suite.
  */
 final class IPTest extends PHPUnit\Framework\TestCase {
+	/**
+	 * Set up.
+	 *
+	 * @before
+	 */
+	public function set_up() {
+		Monkey\setUp();
+
+		Functions\when( 'wp_unslash' )->returnArg();
+	}
+
+	/**
+	 * Tear down.
+	 *
+	 * @after
+	 */
+	public function tear_down() {
+		Monkey\tearDown();
+	}
 
 	/**
 	 * Test `get_ip`.
@@ -23,7 +44,10 @@ final class IPTest extends PHPUnit\Framework\TestCase {
 	 * @param object|null  $trusted_header_data Trusted header data.
 	 */
 	public function test_get_ip( $expect, $server, $trusted_header_data ) {
-		add_test_option( 'trusted_ip_header', $trusted_header_data );
+		Functions\expect( 'get_site_option' )
+			->once()
+			->with( 'trusted_ip_header' )
+			->andReturn( $trusted_header_data );
 
 		$old_server = $_SERVER;
 		$_SERVER    = $server;

```

Reply:
Thanks for your review, @jeherve! I was mainly following the pattern we set out in the WAF package where we have a central `Automattic\Jetpack\Waf` namespace and use the applied structure for testing, but both of these suggestions seem logical to me. 

I can't say at this point whether or not this package will only have one class, but to remain flexible, it's probably better to push for its own namespace rather than potentially limit it. @nateweller what do you think would be appropriate here? Perhaps we keep the `Automattic\Jetpack\IP` namespace but rename the primary class?

Regarding the testing solution, I am not familiar with `brain/monkey` but this certainly seems like a nice approach rather than defining mocked functions every time we come across something we'd like to test that introduces something new. This might also be an excellent approach to consider applying within the WAF package.

Reply:
@jeherve, thanks again - I've applied the suggested testing structure and for now, updated the namespace to `Automattic\Jetpack\Utils\IP` so we can press forward.

Reply:
> Regarding the testing solution, I am not familiar with brain/monkey but this certainly seems like a nice approach rather than defining mocked functions every time we come across something we'd like to test that introduces something new. This might also be an excellent approach to consider applying within the WAF package.

Agreed! +1 from my side for using BrainMonkey in the future

Reply:
Hi @dkmyta.
I think using the `Automattic\Jetpack\Utils` namespace might be a bit confusing, as that would imply there's the `utils` package.
I think @jeherve meant naming the class `Utils`, making it `Automattic\Jetpack\IP\Utils`. I also believe that would make the code better structured and easier to read.

Reply:
@sergeymitr Ah! My mistake - that does make a lot more sense! I've gone ahead and applied those changes 👍🏻